### PR TITLE
Support for inside and outside margins

### LINKED ISSUE #343: Support for inside and outside margins
The `page` function allows for setting the page's `margins`. However, the options are limited to specific sides (`top`, `bottom`, `left`, and `right`) or axis (`horizontal` and `vertical`).

Some templates, like Brazilian's ABNT, have a concept of "outside/outer margins" and "inside/inner margins". Outside margins are the left margin on the left page and the right margin on the right page. Inside margins are in between the two facing pages.

Here is a diagram:

image

I propose these properties to be added to `margin`.

### PR DESCRIPTION
**Pull request by Luis-Licea**
Fixes #343

### CODE REVIEW

File: library/src/layout/page.rs [hunk 1]
```diff
@@ -350,6 +359,16 @@ impl PageElem {
             // The padded width of the page's content without margins.
             let pw = frame.width();
 
+            // Make a clone of the original margin in case you need to modify
+            // the margins for a two-sided document.
+            let mut margin = margin.clone();
```
Comment (PgBiel) [inline library/src/layout/page.rs:364]: Clippy seems to be complaining about this line as `margin` has a type that is `Copy`; does removing `.clone()` still make it work?
```suggestion
            let mut margin = margin;
```

File: library/src/layout/page.rs [hunk 2]
```diff
@@ -87,6 +87,15 @@ pub struct PageElem {
     #[default(false)]
     pub flipped: bool,
 
+    ///When `{true}`, the left and right margins will be swapped on even pages
+    ///to achieve inside and outside margins. This is commonly used when writing
+    ///books, for example. Defaults to `{false}`.
+    /// ```example
+    /// #set page(two-sided: true, margin: (left: 1in, right: 1.25in))
+    /// ```
```
Comment (PgBiel) [inline library/src/layout/page.rs:95]:
```suggestion
    /// When `{true}`, the left and right margins will be swapped on even pages
    /// to achieve inside and outside margins. This is commonly used when writing
    /// books, for example. Defaults to `{false}`.
    ///
    /// ```example
    /// #set page(two-sided: true, margin: (left: 1in, right: 1.25in))
    /// ```
```

File: library/src/layout/page.rs (hunk 1 above)
Comment (Luis-Licea) [reply library/src/layout/page.rs]: Yes. That solved it.

Comment (PgBiel) [reply library/src/layout/page.rs]: Epic. (I can't resolve this thread for some reason, lol)

Please take a look at my other suggestion when you can, regarding spacing in the docs. ;)

File: library/src/layout/page.rs (hunk 2 above)
Comment (Luis-Licea) [reply library/src/layout/page.rs]: I added the leading spaces and re-wrapped the lines to be at maximum 80 characters.

Comment (PgBiel) [reply library/src/layout/page.rs]: Thanks!

**Comment by laurmaedje** [comment]
The implementation looks fine, but I'd prefer an API based on `inside` and `outside` keys in the `margin` dictionary.

**Comment by laurmaedje** [comment]
Marked as draft because of the API design. (Helps me when scanning PRs.)

**Comment by researcher2312** [comment]
Will this PR be merged, or is the work dropped?

**Comment by PgBiel** [comment]
> Will this PR be merged, or is the work dropped?

They faced difficulties in going forward with their contribution, unfortunately. We need someone else to pick it up. I might be able to pick it up in several days if nobody is able to.

**Comment by Luis-Licea** [comment]
The tests always passed for me locally, the PNGs that were produced were the same as the reference PNGs, but manually compiling the PDF gave obviously wrong documents. It should be imossible for the generated PNG and the generated PDF to be different. It felt like the testing framework was broken.

**Comment by Luis-Licea** [comment]
I checked out the commit where I had generated the reference PNG, re-compiled everything, and re-tested everything. The tests somehow passed, but when I used target/debug/typst to generate the very PDF shown in the 'two-sided.png' reference image, the PDF was completely different. So, because the builds were not reproducible, I could not continue deveolpent.

**Comment by PgBiel** [comment]
@Luis-Licea did you compile typst using `cargo build -p typst-cli`?

**Comment by PgBiel** [comment]
That's likely why it was compiling a different PDF; you were likely compiling just the library, which doesn't change the `target/debug/typst` file at all - that file is produced by the `typst-cli` crate (the binary). In other words, you were accidentally using an older version to test your PDFs.

Running `cargo build -p typst-cli` will probably fix your issues.

**Comment by laurmaedje** [comment]
You can also generate the PDFs from the testing process by passing the `--pdf` flag. Though, as you said, your changes should not have any effect on PDF output.

**Comment by Luis-Licea** [comment]
Running `testit --pdf` shows the PDF I wanted to see in `tests/pdf/layout/two-sided.pdf`. So that means the testing is not broken like I thought. However, running `cargo build -p typst-cli` and then compiling `tests/typ/layout/two-sided.typ` produces a completely different PDF.

**Comment by laurmaedje** [comment]
Ah nevermind, that's not even the point. The tests use different defaults for a few things: 120pt x auto pages, 10pt margins and 10pt font size. Compared to A4 paper, 2.5cm margins and 11pt font size. So it will of course look different, nothing to worry about.

**Pull request closed without merging**